pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,420 @@
+			}
+			klog.Infof("Successfully updated PV annotations and tags for volume %s", volume.Name)
+		}
+	}
+	return nil
+}
+
+// updateEBSTags updates the tags of an AWS EBS volume with rate limiting
+func (c *EBSVolumeTagsController) updateEBSTags(ctx context.Context, pvBatch []*v1.PersistentVolume, ec2Client *ec2.EC2,
+	resourceTags []configv1.AWSResourceTag) error {
+	err := c.rateLimiter.Wait(ctx)
+	// If context is cancelled or rate limit cannot be acquired, return error
+	if err != nil {
+		klog.Errorf("Rate limiter error: %v", err)
+		return err
+	}
+
+	// Prepare tags
+	tags := newAndUpdatedTags(resourceTags)
+	// Create or update the tags
+	_, err = ec2Client.CreateTags(&ec2.CreateTagsInput{
+		Resources: pvsToResourceIDs(pvBatch),
+		Tags:      tags,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *EBSVolumeTagsController) listPersistentVolumes() ([]*v1.PersistentVolume, error) {
+	pvList, err := c.commonClient.KubeInformers.InformersFor("").Core().V1().PersistentVolumes().Lister().List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing volumes objects: %v", err)
+		return nil, err
+	}
+	return pvList, nil
+}
+
+func (c *EBSVolumeTagsController) updateVolume(ctx context.Context, pv *v1.PersistentVolume) error {
+	_, err := c.commonClient.KubeClient.CoreV1().PersistentVolumes().Update(ctx, pv, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("error updating volume object %s: %v", pv.Name, err)
+		return err
+	}
+	return nil
+}
+
+func (c *EBSVolumeTagsController) handleTagUpdateFailure(batch []*v1.PersistentVolume, updateErr error) {
+	for _, pv := range batch {
+		klog.Errorf("Error updating volume %v tags: %v", pv.Name, updateErr)
+		c.failedQueue.AddRateLimited(pv.Name)
+	}
+	var pvNames []string
+	for _, pv := range batch {
+		pvNames = append(pvNames, pv.Name)
+	}
+	errorMessage := fmt.Sprintf("Error updating tags for volume %v: %v", pvNames, updateErr)
+	// Emit a warning event for the failure
+	c.eventRecorder.Warning("EBSVolumeTagsUpdateFailed", fmt.Sprintf("Failed to update tags for batch %v: %v", pvNames, errorMessage))
+}
+
+// newAndUpdatedTags adds and update existing AWS tags with new resource tags from OpenShift infrastructure
+func newAndUpdatedTags(resourceTags []configv1.AWSResourceTag) []*ec2.Tag {
+	// Convert map back to slice of ec2.Tag
+	var tags []*ec2.Tag
+	for _, tag := range resourceTags {
+		tags = append(tags, &ec2.Tag{
+			Key:   aws.String(tag.Key),
+			Value: aws.String(tag.Value),
+		})
+	}
+	return tags
+}
+
+func (c *EBSVolumeTagsController) filterUpdatableVolumes(volumes []*v1.PersistentVolume, newTagsHash string) []*v1.PersistentVolume {
+	var updatablePVs []*v1.PersistentVolume
+
+	for _, volume := range volumes {
+		// Check if the volume is a CSI volume with the correct driver
+		if volume.Spec.CSI != nil && volume.Spec.CSI.Driver == driverName &&
+			// Ensure the volume is not already in the failed queue
+			!c.isVolumeInFailedQueue(volume.Name) &&
```

Comment:
Yes @TrilokGeer, Here we cannot update the tags immediately if the volume is in a failed queue.
But it would be even worse if we try to refresh the queue by removing all volume from the queue and then retry to update the tags in batches. as we are calling the sync func here when there is any update on Infrastructure resource.